File: solutions/c/bob/1/bob.c
#include "bob.h"
#include <ctype.h>
#include <stdbool.h>
#include <string.h>

char *hey_bob(char *greeting) {
    const size_t len = strlen(greeting);
    bool isEmpty = true;
    bool isQuestion = false;
    bool hasLower = false;
    bool hasUpper = false;

    for (size_t i = 0; i < len; i++) {
        const char c = greeting[i];
        if (isEmpty && !isspace(c)) {
            isEmpty = false;
        }
        if (c == '?') {
            isQuestion = true;
        }
        if (isalnum(c)) {
            isQuestion = false;
        }
        if (islower(c)) {
            hasLower = true;
        }
        if (isupper(c)) {
            hasUpper = true;
        }
    }
    if (isEmpty) {
        return "Fine. Be that way!";
    }

    const bool isYelling = hasUpper && !hasLower;
    if (isYelling && isQuestion) {
        return "Calm down, I know what I'm doing!";
    }
    if (isYelling) {
        return "Whoa, chill out!";
    }
    if (isQuestion) {
        return "Sure.";
    }
    return "Whatever.";
}

File: solutions/ruby/bob/1/bob.rb
class Bob
  def self.hey(str)
    str = str.strip
    return 'Fine. Be that way!' if str.empty?
    is_yelling = str.match?(/[A-Z]/) && !str.match?(/[a-z]/)
    is_question = str.end_with?('?')
    if is_yelling
      return 'Calm down, I know what I\'m doing!' if is_question
      return 'Whoa, chill out!' 
    end
    return 'Sure.' if is_question
    'Whatever.'
  end
end
